plugins/openpgp-api-lib/openpgp-api/src/main/java/org/openintents/openpgp/OpenPgpApiManager.java

```diff
@@ -123,12 +123,15 @@ public void refreshConnection() {
         }
 
         Intent intent = new Intent(OpenPgpApi.ACTION_CHECK_PERMISSION);
-        getOpenPgpApi().executeApiAsync(intent, null, null, new IOpenPgpCallback() {
-            @Override
-            public void onReturn(Intent result) {
-                onPgpPermissionCheckResult(result);
-            }
-        });
+        OpenPgpApi safeOpenPgpApi = getOpenPgpApi();
+        if (safeOpenPgpApi != null) {
+            safeOpenPgpApi.executeApiAsync(intent, null, null, new IOpenPgpCallback() {
+                @Override
+                public void onReturn(Intent result) {
+                    onPgpPermissionCheckResult(result);
```

Comment:
Thanks for the analysis. I think a simpler fix is to only change the body of this callback to

```java
if (openPgpProviderState != OpenPgpProviderState.UNCONFIGURED) {
    onPgpPermissionCheckResult(result);
}
```

This will avoid doing work that is no longer relevant if `openPgpProviderState` has changed to `UNCONFIGURED` after the request was kicked off.

Reply:
@cketti Just to understand you correctly. Do you mean by "a simpler fix" implementing this part and skip the rest of commit? Or just to reword this chunk only?

Reply:
You are right. If I replace the onReturn callback with your variant while keeping the getOpenPgpApi() part of the commit, the crash is no more triggered.

Force-pushing and +CR from me - ready to merge?

Reply:
Sorry, just realized that the initial stacktrace pointed not to this part of code but to OpenPgpKeyPreference.java.

getOpenPgpApi() may return null object, and in this case your proposed fix is invalid - it will be null.executeAsync throwing another exception.

Reply:
Yes, I meant only do the suggested change and skip the rest. My issue with your fix is that it's fighting symptoms instead of addressing the underlying issue. The problem is that `getOpenPgpApi()` is called when it shouldn't be. We want to avoid getting into that situation instead of trying to get the `OpenPgpApi` instance and then dealing with the fact that it might not be there.

What we could do in addition to the change I suggested, is to add something like the following to `getOpenPgpApi()`:

```java
if (openPgpProviderState == OpenPgpProviderState.UNCONFIGURED) {
    throw new IllegalStateException("Called getOpenPgpApi() while no provider was configured");
}
```

This should help to catch these kinds of errors a bit earlier and provides a bit more context on what went wrong. Unfortunately, it looks like there are a few places that call `getOpenPgpApi()` unconditionally and expect to get `null` back when no provider was configured. But that's a sign that we need to fix the public API of `OpenPgpApiManager` rather than something to embrace.

Reply:
I totally agree that unguarded calls to OpenPGP API functions must be refactored.

More practically, I observed the race condition in the `OpenPgpApiManager`:

1. The user clicks the OpenPgpProvider preference, disabling the provider
2. The handler sets state to UNCONFIGURED but the `refreshConnection` changes UNCONFIGURED back to OK.
3. The `getOpenPgpApi` is called, throwing back `null`
4. The unguarded API call triggers the parcel exchange resulting with a call to non-existent callback.

That's why I extended the `getOpenPgpApi` call to `diconnect` the provider if the preference change is observed. Now the question is, can we guard the broader logic involving `getOpenPgpApi` call? If not, we cannot solve the issue other than embracing the `null` value.

If you want, I can make a debug build for unmodified snapshot and attach a debug log appended with procedure calls with Timber.

Reply:
Also I think we must invite @Valodim as he was the maintainer of `OpenPgpApiManager`